core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/TrainPhysicsIntegrator.java

```diff
@@ -65,39 +85,66 @@ private IntegrationStep step(double timeStep, double initialLocation, double ini
+                computeAcceleration(rollingStock, rollingResistance, weightForce, speed, tractionForce, directionSign);
         return newtonStep(timeStep, speed, acceleration, directionSign);
     }
 
-    /** Compute the weight force of a rolling stock at a given position on a given path */
-    public static double getWeightForce(PhysicsRollingStock rollingStock, PhysicsPath path, double headPosition) {
+    private double getDeceleration(double speed, double position) {
+        assert (action == Action.BRAKE);
+        if (brakingType == BrakingType.CONSTANT) return rollingStock.getDeceleration();
+        var grade = getMinGrade(rollingStock, path, position);
+        var mRotating = grade >= 0 ? mRotatingMax : mRotatingMin;
+        var gradientAcceleration = getGradientAcceleration(grade, mRotating);
+        return switch (brakingType) {
+            case ETCS_EBD -> -rollingStock.getSafeBrakingAcceleration(speed) + gradientAcceleration;
+            case ETCS_SBD -> -rollingStock.getServiceBrakingAcceleration(speed) + gradientAcceleration;
+            case ETCS_GUI -> -rollingStock.getNormalServiceBrakingAcceleration(speed)
+                    + gradientAcceleration
+                    + rollingStock.getGradientAccelerationCorrection(gradientAcceleration, speed);
+            default -> throw new UnsupportedOperationException("Braking type not supported: " + brakingType);
+        };
+    }
+
+    /** Compute the average grade of a rolling stock at a given position on a given path in m/km */
+    public static double getAverageGrade(PhysicsRollingStock rollingStock, PhysicsPath path, double headPosition) {
         var tailPosition = Math.min(Math.max(0, headPosition - rollingStock.getLength()), path.getLength());
         headPosition = Math.min(Math.max(0, headPosition), path.getLength());
-        var averageGrade = path.getAverageGrade(tailPosition, headPosition);
+        return path.getAverageGrade(tailPosition, headPosition);
+    }
+
+    /** Compute the weight force of a rolling stock at a given position on a given path */
+    public static double getWeightForce(PhysicsRollingStock rollingStock, double grade) {
         // get an angle from a meter per km elevation difference
         // the curve's radius is taken into account in meanTrainGrade
-        var angle = Math.atan(averageGrade / 1000.0); // from m/km to m/m
+        var angle = Math.atan(grade / 1000.0); // from m/km to m/m
         return -rollingStock.getMass() * 9.81 * Math.sin(angle);
     }
 
+    /** Compute the min grade of a rolling stock at a given position on a given path in m/km */
+    public static double getMinGrade(PhysicsRollingStock rollingStock, PhysicsPath path, double headPosition) {
+        var tailPosition = Math.min(Math.max(0, headPosition - rollingStock.getLength()), path.getLength());
+        headPosition = Math.min(Math.max(0, headPosition), path.getLength());
+        return path.getMinGrade(tailPosition, headPosition);
```

Comment:
Oh, if we call it at *every integration step* we may need the optimized version sooner than I thought. Though that can still wait until after the first functional POC.